packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js

```diff
@@ -41,333 +32,320 @@ const DEFAULT_QUERY = {
+			term.name.toLowerCase() === name.toLowerCase()
+		);
+	} );
+}
+
+/**
+ * Get filter matcher function.
+ *
+ * @param {string} filterValue Filter value.
+ * @return {(function(Object): (Object|boolean))} Matcher function.
+ */
+function getFilterMatcher( filterValue ) {
+	const matchTermsForFilter = ( originalTerm ) => {
+		if ( '' === filterValue ) {
+			return originalTerm;
+		}
+
+		// Shallow clone, because we'll be filtering the term's children and
+		// don't want to modify the original term.
+		const term = { ...originalTerm };
+
+		// Map and filter the children, recursive so we deal with grandchildren
+		// and any deeper levels.
+		if ( term.children.length > 0 ) {
+			term.children = term.children
+				.map( matchTermsForFilter )
+				.filter( ( child ) => child );
+		}
+
+		// If the term's name contains the filterValue, or it has children
+		// (i.e. some child matched at some point in the tree) then return it.
+		if (
+			-1 !==
+				term.name.toLowerCase().indexOf( filterValue.toLowerCase() ) ||
+			term.children.length > 0
+		) {
+			return term;
+		}
+
+		// Otherwise, return false. After mapping, the list of terms will need
+		// to have false values filtered out.
+		return false;
+	};
+	return matchTermsForFilter;
+}
+
+/**
+ * Hierarchical term selector.
+ *
+ * @param {Object} props      Component props.
+ * @param {string} props.slug Taxonomy slug.
+ * @return {WPElement}        Hierarchical term selector component.
+ */
+function HierarchicalTermSelector( { slug } ) {
+	const instanceId = useInstanceId( HierarchicalTermSelector );
+	const [ adding, setAdding ] = useState( false );
+	const [ formName, setFormName ] = useState( '' );
+	/**
+	 * @type {[number|'', Function]}
+	 */
+	const [ formParent, setFormParent ] = useState( '' );
+	const [ showForm, setShowForm ] = useState( false );
+	const [ filterValue, setFilterValue ] = useState( '' );
+	const [ filteredTermsTree, setFilteredTermsTree ] = useState( [] );
+	const debouncedSpeak = useDebounce( speak, 500 );
+
+	const selector = ( select ) => {
+		const { getCurrentPost } = select( editorStore );
+		const { getTaxonomy, getEntityRecords, isResolving } = select(
+			coreStore
+		);
+		const taxonomy = getTaxonomy( slug );
+
+		const availableTerms =
+			getEntityRecords( 'taxonomy', slug, DEFAULT_QUERY ) || [];
+
+		const loading = isResolving( 'getEntityRecords', [
+			'taxonomy',
+			slug,
+			DEFAULT_QUERY,
+		] );
+
+		const terms = taxonomy
+			? select( editorStore ).getEditedPostAttribute( taxonomy.rest_base )
+			: [];
+
+		return {
+			hasCreateAction: taxonomy
+				? get(
+						getCurrentPost(),
+						[ '_links', 'wp:action-create-' + taxonomy.rest_base ],
+						false
+				  )
+				: false,
+			hasAssignAction: taxonomy
+				? get(
+						getCurrentPost(),
+						[ '_links', 'wp:action-assign-' + taxonomy.rest_base ],
+						false
+				  )
+				: false,
+			terms,
+			loading,
+			availableTerms,
+			availableTermsTree: sortBySelected(
```

Comment:
This will generate a new array/object on each selector causing this component to rerender each time there's a change (even an unrelated one like typing in a block), you should consider moving this out of the selector and potentially use `useMemo` to compute it.